File: event-consumer/src/entities/Metric.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  Index
} from "typeorm";

@Entity("metrics")
export class Metric {
  @PrimaryGeneratedColumn("uuid")
  id!: string;

  @Column({ type: "varchar", length: 50 })
  @Index()
  service!: string;

  @Column({ type: "uuid", nullable: true })  // ✅ Added traceId
  @Index()
  traceId!: string;

  @Column({ type: "varchar", length: 10, nullable: true })  // ✅ Added method
  method!: string;

  @Column({ type: "varchar", length: 255, nullable: true })  // ✅ Added path
  path!: string;

  @Column({ type: "timestamp" })
  @Index()
  timestamp!: Date;

  @Column({ type: "float" })
  responseTimeMs!: number;

  @Column({ type: "int" })
  statusCode!: number;

  @Column({ type: "int" })
  requestCount!: number;

  @Column({ type: "int" })
  errorCount!: number;

  @Column({ type: "int", nullable: true })
  responseSizeBytes!: number;

  @CreateDateColumn()
  createdAt!: Date;
}

File: ml-service/app/schemas/metric.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional

class MetricData(BaseModel):
    id: str
    service: str
    trace_id: Optional[str] = None
    method: Optional[str] = None
    path: Optional[str] = None
    timestamp: datetime
    response_time_ms: float
    status_code: int
    request_count: int
    error_count: int
    response_size_bytes: Optional[int] = None
    created_at: datetime

class AnomalyAlert(BaseModel):
    alert_id: str
    timestamp: datetime
    service: str
    trace_id: Optional[str] = None
    method: Optional[str] = None
    path: Optional[str] = None
    metric_id: str
    anomaly_score: float
    anomaly_type: str
    details: dict
    threshold: float
